include/cadmium/modeling/coupling.hpp

```diff
@@ -0,0 +1,148 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef CADMIUM_COUPLING_HPP
+#define CADMIUM_COUPLING_HPP
+
+#include<type_traits>
+#include<tuple>
+
+namespace cadmium::modeling {
+
+    /**
+     * model_tuple is a struct used to keep the TIME as template parameter for the models being coupled
+     */
+    template<template<typename TIME> class... Ms>
+    struct models_tuple {
+        template<typename T>
+        using type=std::tuple<Ms<T>...>;
+    };
+
+    /**
+     * EIC is a coupling from an input port in the coupled model to an input port of a sub model
+     */
+    template<typename EXTERNAL_PORT, template<typename TIME> class SUBMODEL, typename SUBMODEL_PORT>
+    struct EIC {
+        using external_input_port=EXTERNAL_PORT;
+        template<typename TIME>
+        using submodel=SUBMODEL<TIME>;
+        using submodel_input_port=SUBMODEL_PORT;
+    };
+
+    /**
+     * EOC is a coupling from a submodel output port to an outpit port in the coupled model
+     */
+    template<template<typename TIME> class SUBMODEL, typename SUBMODEL_PORT, typename EXTERNAL_PORT>
+    struct EOC {
+        template<typename TIME>
+        using submodel=SUBMODEL<TIME>;
+        using submodel_output_port=SUBMODEL_PORT;
+        using external_output_port=EXTERNAL_PORT;
+    };
+
+    /**
+     * IC is a coupling from a submodel output port to an output port in the coupled model
+     */
+    template<template<typename TIME> class SUBMODEL_FROM, typename PORT_FROM,
+            template<typename TIME> class SUBMODEL_TO, typename PORT_TO>
+    struct IC {
+        template<typename TIME>
+        using from_model=SUBMODEL_FROM<TIME>;
+        using from_model_output_port=PORT_FROM;
+        template<typename TIME>
+        using to_model=SUBMODEL_TO<TIME>;
+        using to_model_input_port=PORT_TO;
+    };
+
+
+    namespace pdevs {
+        /**
+         * A couple model is a list of submodels, input ports, output ports and couplings EICs, ICs, EOCs.
+         * Template parameter TIME is not required for coupled models, because they do not produce behavior
+         * since coupled model are structural definitions only.
+         *
+         * IP are a tuple of input_port
+         * OP are a tuple of output_port
+         * MODELS a models_tuple (the struct is a helper for simplifying the definition of tuple of templates)
+         * EICs is a tuple of EIC coupling input ports of the coupled model to input ports in submodels
+         * EOCs is a tuple of EOC coupling output ports in submodels to output ports of the coupled model
+         * ICs is a tuple IC coupling output ports in submodels to input ports in submodels
+         * SELECT is an optional template class with an imminent function that, given a list of imminent candidate
```

Comment:
What is doing this SELECT docstring in the pdevs coupled_model structure?

Reply:
Fixed, those come from an error in a commit. Solved. 
About the doc format, I think we should discuss that one offline and when deciding which way to go fix it in a different PR.